### PR TITLE
[prometheus-to-sd] Dynamic sources

### PR DESCRIPTION
**Pull request by Szetty**
First PR to make prometheus-to-sd be able to work as a daemon set.
In this version the pods are taken only at the start, but in the next versions they will be "watched" and updated.

### CODE REVIEW

**Comment by Szetty** [comment]
cc @loburm @serathius

File: prometheus-to-sd/config/api_server_client.go [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+		}
+	}
+	return flags.Uri{}
+}
+
+func mapToSourceConfig(uri flags.Uri, ip string, podConfig PodConfig) (*SourceConfig, error) {
+	componentName := uri.Key
+	port := uri.Val.Port()
+	whitelisted := uri.Val.Query().Get("whitelisted")
+	return newSourceConfig(componentName, ip, port, defaultMetricsPath, whitelisted, podConfig)
```
[the remark this answers is not in the document]
Comment (Szetty) [reply prometheus-to-sd/config/api_server_client.go]: this is how it is for static source config too

File: prometheus-to-sd/config/api_server_client_test.go [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+	"net/url"
+	"testing"
+)
+
+func TestMapToSourceConfig(t *testing.T) {
+	testcases := []struct {
+		uri       flags.Uri
+		ip        string
+		podConfig PodConfig
+		expected  SourceConfig
```
[the remark this answers is not in the document]
Comment (Szetty) [reply prometheus-to-sd/config/api_server_client_test.go]: done

File: prometheus-to-sd/config/api_server_client.go [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+	if err != nil {
+		return nil, err
+	}
+	return clientSet, nil
+}
+
+func createLabelSelector(sources []flags.Uri) string {
+	var nameList string
+	for _, sourceUri := range sources {
+		if sourceUri.Key != "" {
```
[the remark this answers is not in the document]
Comment (Szetty) [reply prometheus-to-sd/config/api_server_client.go]: done

File: prometheus-to-sd/config/api_server_client_test.go [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+		if assert.NoError(t, err) {
+			assert.Equal(t, testcase.expected, *output)
+		}
+	}
+}
+
+func TestCreateLabelSelector(t *testing.T) {
+	testcases := []struct {
+		input    []flags.Uri
+		expected string
```
[the remark this answers is not in the document]
Comment (Szetty) [reply prometheus-to-sd/config/api_server_client_test.go]: done

File: prometheus-to-sd/config/api_server_client.go [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+)
+
+const (
+	podNamespace = "kube-system"
+	nameLabel    = "k8s-app"
+)
+
+// SourceConfigsFromApiServer takes pod specifications from the Kubernetes API and maps them to source configs
```
[the remark this answers is not in the document]
Comment (Szetty) [reply prometheus-to-sd/config/api_server_client.go]: done

**Comment by loburm** [comment]
/lgtm
/approve
But please wait for comments from @serathius

File: prometheus-to-sd/config/api_server_client.go [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+		return nil, err
+	}
+	podResponse, err := kubeApi.CoreV1().Pods(podNamespace).List(v1.ListOptions{
+		FieldSelector: fmt.Sprintf("spec.nodeName=%s", gceConfig.Instance),
+		LabelSelector: createLabelSelector(sources),
+	})
+	if err != nil {
+		return nil, err
+	}
+	pods := podResponse.Items
```
[the remark this answers is not in the document]
Comment (Szetty) [reply prometheus-to-sd/config/api_server_client.go]: done

File: prometheus-to-sd/config/api_server_client.go [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+	return getConfigsFromPods(pods, sources), nil
+}
+
+func setupKubernetesApi() (clientset.Interface, error) {
```
[the remark this answers is not in the document]
Comment (Szetty) [reply prometheus-to-sd/config/api_server_client.go]: done

File: prometheus-to-sd/config/api_server_client.go [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+	conf, err := rest.InClusterConfig()
+	if err != nil {
+		return nil, err
+	}
+	clientSet, err := clientset.NewForConfig(conf)
+	if err != nil {
```
[the remark this answers is not in the document]
Comment (Szetty) [reply prometheus-to-sd/config/api_server_client.go]: done

File: prometheus-to-sd/config/api_server_client_test.go [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+
+func TestMapToSourceConfig(t *testing.T) {
+	testcases := []struct {
+		uri       flags.Uri
+		ip        string
+		podConfig PodConfig
+		want      SourceConfig
+	}{
+		{
+			flags.Uri{Key: "kube-proxy", Val: url.URL{Host: ":8080"}},
```
[the remark this answers is not in the document]
Comment (Szetty) [reply prometheus-to-sd/config/api_server_client_test.go]: done

File: prometheus-to-sd/config/api_server_client_test.go [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+			"10.25.78.143",
+			PodConfig{
+				NamespaceId: "kube-system",
+				PodId:       "something",
+			},
+			SourceConfig{
+				"kube-proxy",
```
[the remark this answers is not in the document]
Comment (Szetty) [reply prometheus-to-sd/config/api_server_client_test.go]: done

File: prometheus-to-sd/config/api_server_client_test.go [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+}
+
+func TestCreateLabelSelector(t *testing.T) {
+	testcases := []struct {
+		input []flags.Uri
+		want  string
+	}{
+		{
+			[]flags.Uri{},
+			"k8s-app in ()",
```
[the remark this answers is not in the document]
Comment (Szetty) [reply prometheus-to-sd/config/api_server_client_test.go]: these all are correct queries, I had incorrect queries too, but I thought about this too

File: prometheus-to-sd/main.go [hunk 12]
```diff
@@ -111,6 +116,28 @@ func main() {
 	<-make(chan int)
 }
 
+func validateDynamicSources(sources flags.Uris) {
```
[the remark this answers is not in the document]
Comment (Szetty) [reply prometheus-to-sd/main.go]: done

File: prometheus-to-sd/config/api_server_client.go [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+		nameList += sourceUri.Key + ","
+	}
+	return fmt.Sprintf("%s in (%s)", nameLabel, nameList)
+}
+
+func getConfigsFromPods(pods []core.Pod, sources []flags.Uri) []SourceConfig {
+	var sourceConfigs []SourceConfig
+	for _, pod := range pods {
+		source := findSourceWithKey(sources, pod.Labels[nameLabel])
+		ip := pod.Status.PodIP
```
[the remark this answers is not in the document]
Comment (Szetty) [reply prometheus-to-sd/config/api_server_client.go]: done

File: prometheus-to-sd/config/api_server_client.go [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+	return sourceConfigs
+}
+
+func findSourceWithKey(sources []flags.Uri, host string) flags.Uri {
+	for _, source := range sources {
+		if source.Key == host {
+			return source
+		}
+	}
+	return flags.Uri{}
```
[the remark this answers is not in the document]
Comment (Szetty) [reply prometheus-to-sd/config/api_server_client.go]: done

File: prometheus-to-sd/main.go [hunk 15]
```diff
@@ -61,26 +61,31 @@ var (
...
+	debugPort      = flag.Uint("port", 6061, "Port on which debug information is exposed.")
+	dynamicSources = flags.Uris{}
 
 	customMetricsPrefix = "custom.googleapis.com"
 )
 
 func main() {
 	flag.Set("logtostderr", "true")
 	flag.Var(&source, "source", "source(s) to watch in [component-name]:http://host:port?whitelisted=a,b,c format")
+	flag.Var(&dynamicSources, "dynamic-source", "dynamic source(s) to watch in format: `[component-name]:http://:port?whitelisted=metric1,metric2`")
```
[the remark this answers is not in the document]
Comment (Szetty) [reply prometheus-to-sd/main.go]: done

File: prometheus-to-sd/config/api_server_client.go [hunk 16]
```diff
@@ -0,0 +1,91 @@
...
+		sourceConfig, err := mapToSourceConfig(source, ip, podConfig)
+		if err != nil {
+			glog.Warning("could not create source config for pod %s: %v", pod.Name, err)
+		}
+		sourceConfigs = append(sourceConfigs, *sourceConfig)
+	}
+	return sourceConfigs
+}
+
+func findSourceWithKey(sources []flags.Uri, host string) flags.Uri {
```
[the remark this answers is not in the document]
Comment (Szetty) [reply prometheus-to-sd/config/api_server_client.go]: we have one pod per source not one source per pod, that should be okay

File: prometheus-to-sd/config/api_server_client.go [hunk 17]
```diff
@@ -0,0 +1,104 @@
...
+		if err != nil {
+			glog.Warning("could not create source config for pod %s: %v", pod.Name, err)
+		}
+		sourceConfigs = append(sourceConfigs, *sourceConfig)
+	}
+	return sourceConfigs
+}
+
+func findSourceWithKey(sources []flags.Uri, key string) (flags.Uri, error) {
+	for _, source := range sources {
```
Comment (serathius) [inline prometheus-to-sd/config/api_server_client.go]: Use map[string]url for mapping on components

**Comment by serathius** [comment]
Please change name `api_server_client.go` to `dynamic_source.go`

File: prometheus-to-sd/config/api_server_client.go (hunk 17 above)
Comment (serathius) [inline prometheus-to-sd/config/api_server_client.go]: Use map[string]url for mapping on components

**Pull request merged**